Verify pointers during functable init

Dead2 opened the pull request:
During init, make sure none of the functable entries are nullpointers.
This will cause zlib-ng to abort during init if it is run on a cpu that is missing both optimized and generic fallback functions, instead of running fine until such a missing function is called and potentially being missed during testing.

Dead2 committed:
Improve resilience of the functable initialization; during functable init,
make sure none of the function pointers are nullpointers.

Up until now, zlib-ng and the application would have segfaulted either at the start
of processing, or at some point later depending on when a nullpointer call would happen
in the processing. In any case most likely after accepting data from the application.

Now, the deflateinit/inflateinit functions will error with Z_VERSION_ERROR, and
gzopen will return Z_STREAM_ERROR before actually processing any data.

Direct calls to functions like adler32 or crc32 will however print an error message
and call abort(), as these functions have no actual way of reporting errors.

Note: This should never happen with default builds of zlib-ng, only if it is run on
a cpu that is missing both the matching optimized and the generic fallback functions.
This can currently only happen if zlib-ng is compiled using custom cflags or by
editing the code.

nmoinvaz commented on functable.c:
```diff
@@ -38,6 +38,14 @@
+        fprintf(stderr, "Functable entry not set!\n"); \
+        abort(); \
+    } \
+    FUNCTABLE_ASSIGN(VAR, FUNC_NAME);
```
> Improve error message to identify the missing function.

This is a nice idea.

Dead2 replied:
@nmoinvaz I agree it would be convenient for debugging problems in that area.
But since the name would be hardcoded into the fprintf call, that whole branch segment would be duplicated for every function name instead of it sharing the same code segment.

I am not sure it is justifiable to add that for something that is not really useful for the end user and also _should_ never happen. (I don't think we have had any bugreports about a crash caused by this so far).

mtl1979 replied:
> @nmoinvaz I agree it would be convenient for debugging problems in that area. But since the name would be hardcoded into the fprintf call, that whole branch segment would be duplicated for every function name instead of it sharing the same code segment.
> 

As string literals are practically `const char *`, you could add the preprocessor token value as a variable before calling fprintf and it would be able to deduplicate the code inside the `if` block.

nmoinvaz replied:
@Dead2 yeah that is a good point.

Dead2 commented:
I figured calling `abort()` would not be very nice to the application linking zlib-ng.
Implemented improved failure handling for deflateinit and inflateinit.

Suggestions for how to handle other exported functions directly? Currently I guess they'll likely segfault, same as they would without this PR.
Edit: Or they get infinite recursion like coderabbitai mentions. Its "Solution 2" might be ok for those calling functions directly..?

mtl1979 commented:
> I figured calling `abort()` would not be very nice to the application linking zlib-ng. Implemented improved failure handling for deflateinit and inflateinit.
> 
> Suggestions for how to handle other exported functions directly? Currently I guess they'll likely segfault, same as they would without this PR. Edit: Or they get infinite recursion like coderabbitai mentions. Its "Solution 2" might be ok for those calling functions directly..?

Aborting with error message is still better than "Segmentation Fault" without any clue what went wrong... Solution 2 would most likely result in hidden data corruption as not all functions return status code.

Dead2 commented:
> Aborting with error message is still better than "Segmentation Fault" without any clue what went wrong... 

Yeah, I agree.

> Solution 2 would most likely result in hidden data corruption as not all functions return status code.

No, solution 2 is doing printf + abort for all the direct-calls, while deflateinit/inflateinit would still return Z_VERSION_ERROR (I figure that is the error message most similar, in this case indicating that the zlib version tring to run is not compiled to be compatible with the current cpu).
Inflateinit/deflateinit would never complete, and zlib-ng would never even get to a state of accepting data, so no data loss there. 
Potential data loss for the direct calls that hit the abort() though, but as we agree, that is effectively no worse than a segfault anyways, and I don't see any other proper way to do it.

Again, this is an error condition that should never occur with a default build, only potentially possible with custom build configs where generic functions are left out as well as the needed optimized functions. Currently should not be possible with CMake/Configure without using custom defines or code changes. Even with #1984, this is not possible since it makes sure to require WITH_SSE2 and WITH_OPTIM before allowing the removal of the specific generic fallbacks that SSE2 can replace.

But I feel that it is better to have a defined failure mode IF this ever happens.

I'll work on a solution for the stub functions. I also think several of the stub functions should be unreachable (not directly callable and always after inflateinit/deflateinit), not sure whether we should keep those as functional unreachable stubs or replace them with a common one that just calls abort (not something for this PR in any case).

Edit: Implemented changes and squashed the commits, also improved commit message.

Dead2 commented on functable.c:
```diff
@@ -38,18 +38,26 @@
+        fprintf(stderr, "Functable entry not set, function(s) missing!\n"); \
+        return 1; \
+    } \
+    FUNCTABLE_ASSIGN(VAR, FUNC_NAME);
```
Implemented a variant of Solution 2.

mtl1979 commented:
> I'll work on a solution for the stub functions. I also think several of the stub functions should be unreachable (not directly callable and always after inflateinit/deflateinit), not sure whether we should keep those as functional unreachable stubs or replace them with a common one that just calls abort (not something for this PR in any case).

A lot of changes have been made with the stub functions since I initially designed the logic... I haven't been happy with all the changes and have already said so earlier... It might be best to do proper redesign/rewrite and not just commit random patches that cause more dead code. In the current state, some of the stub functions can be called, but others are essentially no-op in normal code. Coverity scan will get very low coverage due to stub calls being mutually exclusive.
